### PR TITLE
Predictable generation of import library on Windows.

### LINKED ISSUE #4561: Building Windows DLL without exports fails
### Description of the problem / feature request:

In order to produce import library for a DLL bazel uses `/IMPLIB` flag of the linker and expect that it will force creating import library in the place specified:
https://msdn.microsoft.com/en-us/library/67wc07b9.aspx

Unfortunately, this flag is not managing the fact of creating or not creating import library but only sets the name for it. If DLL does not export anything, linker do not create it by design. Unluckily, we actively use DLLs without imports for a reason.

The output of bazel is as follows:
```
C:\Users\Chebotarev_V\Documents\bazel-issues\windows-dll-without-exports>bazel build ...
INFO: Analysed target //:hellolib.dll (8 packages loaded).
INFO: Found 1 target...
ERROR: C:/users/chebotarev_v/documents/bazel-issues/windows-dll-without-exports/BUILD:3:1: output 'hellolib.ifso' was not created
ERROR: C:/users/chebotarev_v/documents/bazel-issues/windows-dll-without-exports/BUILD:3:1: not all outputs were created or valid
Target //:hellolib.dll failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 49.652s, Critical Path: 1.14s
FAILED: Build did NOT complete successfully
```

Please find an example in my repository:
https://github.com/excitoon/bazel-issues/tree/master/windows-dll-without-exports

### Feature requests: what underlying problem are you trying to solve with this feature?

We need building DLLs without exports.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

https://github.com/excitoon/bazel-issues/tree/master/windows-dll-without-exports

### What operating system are you running Bazel on?

Windows 10 x64

### What's the output of `bazel info release`?

```
Build label: 0.10.0
Build target: bazel-out/x64_windows-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Thu Nov 9 22:42:19 +50056 (1517474731339)
Build timestamp: 1517474731339
Build timestamp as int: 1517474731339
```

###  Have you found anything relevant by searching the web?

Nothing expect mentioned.

### PR DESCRIPTION
**Pull request by excitoon**
Fixes #4561 .

### CODE REVIEW

**Comment by excitoon** [comment]
@meteorclody Please kindly look at `testGetDefFileOfSharedLibraryFromCcBinary`. That is most breaking aspect of my change.

As for other tests, I guess, I've accidentally made Moebius strip here (we now need def parser to build def parser), I should not have used def parser at all at least. Funny thing is that this error was not handled by Bazel.

**Comment by excitoon** [comment]
I also wonder why `def_parser` is named so, given that it parses object files and generates `def`.

**Comment by meteorcloudy** [comment]
@excitoon Thanks for this contribution! I named it as `def_parser` exactly because it generates DEF file. Sorry if it's a bit confusion, I was just finding it a bit difficult to keep the name short and precise.

File: src/main/java/com/google/devtools/build/lib/rules/cpp/CcBinary.java [hunk 1]
```diff
@@ -399,18 +399,21 @@ public static ConfiguredTarget init(CppSemantics semantics, RuleContext ruleCont
...
-              || (library.getDynamicLibrary() == null && library.getInterfaceLibrary() == null)) {
-            if (library.getPicStaticLibrary() != null) {
-              if (library.getPicObjectFiles() != null) {
-                objectFiles.addAll(library.getPicObjectFiles());
-              }
-            } else if (library.getStaticLibrary() != null) {
-              if (library.getObjectFiles() != null) {
-                objectFiles.addAll(library.getObjectFiles());
+
+        if (CppHelper.shouldExportAllSymbols(featureConfiguration)) {
```
[the remark this answers is not in the document]
Comment (excitoon) [reply src/main/java/com/google/devtools/build/lib/rules/cpp/CcBinary.java]: I guess the cause of `NullPointerException` here is using generated DEF file in `def_parser`. That makes `def_parser` to depend in `def_parser`. I shall write independent action which would output `LIBRARY nnn` line to DEF file.

[the remark this answers is not in the document]
Comment (meteorcloudy) [reply src/main/java/com/google/devtools/build/lib/rules/cpp/CcBinary.java]: Ah, sounds right to me.

[the remark this answers is not in the document]
Comment (meteorcloudy) [reply src/main/java/com/google/devtools/build/lib/rules/cpp/CcBinary.java]: https://github.com/bazelbuild/bazel/blob/1f684e1b87cd8881a0a4b33e86ba66743e32d674/src/main/java/com/google/devtools/build/lib/bazel/rules/cpp/BazelCppRuleClasses.java#L271 explains why Bazel didn't catch the cycle dependency issue. We're actually explicitly not adding def_parser as a dependency of itself.

File: src/main/java/com/google/devtools/build/lib/rules/cpp/CcLibrary.java [hunk 2]
```diff
@@ -312,15 +325,6 @@ public static void init(
             throw new IllegalStateException("Should not be reached");
           }
         }
-
-        // If user specifies a custom DEF file, then we use this one instead of the generated one.
-        Artifact customDefFile = null;
-        if (ruleContext.isAttrDefined("win_def_file", LABEL)) {
```
Comment (excitoon) [inline src/main/java/com/google/devtools/build/lib/rules/cpp/CcLibrary.java]: This check was redundant.

**Comment by excitoon** [comment]
@meteorcloudy I'm done with changes. The only thing I don't get is why in `testGetDefFileOfSharedLibraryFromCcBinary` even without `windows_exports_all_symbols` we check that all symbols are in DEF file. After my change, in that case DEF file has no symbols.

**Comment by excitoon** [comment]
I've just looked into `def_parser` and found that it does actually parse DEF files (I guess for transitive gathering of symbols?), but currently this mechanism is not used because only object files are being passed through it:
https://github.com/bazelbuild/bazel/blame/master/src/main/java/com/google/devtools/build/lib/rules/cpp/CppHelper.java#L806

**Comment by meteorcloudy** [comment]
> The only thing I don't get is why in testGetDefFileOfSharedLibraryFromCcBinary even without windows_exports_all_symbols we've checked that all symbols are in DEF file.

Previously, the DEF file is always generated by parsing the object files, you can get the file by `def_file` output group. But without `windows_exports_all_symbols` feature it won't be passed to the linker. 

This behavior is actually used at 
https://github.com/tensorflow/tensorflow/blob/38fd4259e1febcb59ea51dea1d2eed2c43ed80e3/tensorflow/BUILD#L586-L590

Do you think you can keep the original behavior?

**Comment by meteorcloudy** [comment]
> I've just looked into `def_parser` and found that it does actually parse DEF files (I guess for transitive gathering of symbols?), but currently this mechanism is not used because only object files are being passed through it:
> https://github.com/bazelbuild/bazel/blame/master/src/main/java/com/google/devtools/build/lib/rules/cpp/CppHelper.java#L806

That's correct.

**Comment by excitoon** [comment]
@meteorcloudy Is it possible that they just add `windows_export_all_symbols` feature for these targets? I can keep original behavior if I add second DEF file which would only exist if `windows_export_all_symbols` is disabled. But I don't see it as a clean solution.

**Comment by meteorcloudy** [comment]
@excitoon OK, if adding `windows_export_all_symbols` gives the DEF file containing parsed symbols, I'm fine with that.

**Comment by excitoon** [comment]
I guess I shall make it as an incompatible change fix.

**Comment by meteorcloudy** [comment]
@excitoon Incompatible change sounds good, but this feature is only being used for TensorFlow on Windows, I would be surprised if anyone else is using it. I can easily send a change to add the `windows_export_all_symbols` feature. So I'm also ok with not going through the incompatible flags process.

**Comment by excitoon** [comment]
@meteorcloudy I'm perfectly fine with not going through this process. Not blocking it anymore.

File: src/test/py/bazel/bazel_windows_cpp_test.py [hunk 3]
```diff
@@ -432,9 +445,9 @@ def testGetDefFileOfSharedLibraryFromCcBinary(self):
     # we should still be able to get the DEF file by def_file output group.
     def_file = os.path.join(bazel_bin, 'main/main.dll.gen.def')
     self.assertTrue(os.path.exists(def_file))
-    self.AssertFileContentContains(def_file, 'hello_A')
-    self.AssertFileContentContains(def_file, 'hello_B')
-    self.AssertFileContentContains(def_file, 'hello_C')
```
Comment (meteorcloudy) [inline src/test/py/bazel/bazel_windows_cpp_test.py]: Can you duplicate this test with `windows_export_all_symbols`?

**Comment by meteorcloudy** [comment]
@excitoon this PR seems to be very useful for solving https://github.com/bazelbuild/bazel/issues/4561, do you have time to get it merged?

**Comment by aiuto** [comment]
ping

**Comment by meteorcloudy** [comment]
https://github.com/bazelbuild/bazel/pull/9976 is a replacement of this PR. This one can be closed.

**Pull request closed without merging**